### PR TITLE
Major refactor

### PR DESCRIPTION
**Pull request by nistath**
This is a major refactor of the CANlib and the generator, with a few goals:
- Make sense for multiple buses
- Be future compatible (for multiplexing mainly)
- Allow more ubiquitous use of packing and unpacking functions thus promoting architecture agnostic usage
  Currently the Teensy and STM have much more capable CAN libraries that can still benefit from spec standardization without the use of a driver.

This refactor is a middle path -- not yet eradicating drivers.

## Changes
- Got rid of boards concept -- we'll either use it properly or not include it at all
- Formalized bus and message concepts
- Refactored generator
  - Changed all strings to single quotes to avoid inconsistencies
  - Made strings satisfy my ocd wherever possible
  - Extended use of str.format
  - Improved multi-file performance by passing already parsed spec in write functions
  - Added shorthand for f.write
  - I want to move all generator files in their own directory, but can't bother with Python packages yet
  - Bunch of stuff I may have forgotten

It's helpful to read `common.py` first.

## Buses, messages, and segments
`constants.h` includes definitions for message **attributes**, such as can_id and frequency.
Those are called by `coord(bus.name, msg.name, attrnm)`

`Can_Library.h` includes definitions for message **forms**, enumerating the different forms that can be received. There can be a one to many mapping between a CAN message identified by its can_id and a message form -- allowing for concepts such as multiplexing.
Those are called by `coord(bus.name, msg.name)`.

All message forms have their own struct, defined in `structs.c`.
Those structs are called by `'{}_T'.format(coord(bus.name, msg.name))`.

Segments still get their own enums and are their own types in `enum_segments.c`, just with better naming.
Those types are called by `'{}_T'.format(coord(bus.name, msg.name, seg.name))` and the included enum values by `coord(bus.name, msg.name, seg.name, val.name)`. It would be a problem is someone had a value named `T` -- this nuisance exists in many other places -- we'll avoid it manually via spec for now.

## Generator
In `common.py`, the `coord` function is an attempt to standardize naming. Ideally there would exist functions that take a bus[->message[->segment]] coordinate and return the proper name. This would avoid confusion between multiple files needing to name things based on the same concept.

### CODE REVIEW

File: generator/Can_Library_h.py [hunk 1]
```diff
@@ -1,75 +1,72 @@
...
 
-        # Include static stuff
-        f.write('#include "static_can.h"\n\n')
+            fw('} ' + '{}_T;\n\n'.format(coord(bus.name)))
 
-        # Add board includes
-        for board in car.boards:
-            f.write('#include "boards/' + board.name + '.h"\n')
-        f.write("\n")
+            fw('{0}_T {0}_IDENTIFY(Frame* frame)'.format(coord(bus.name)) + '\n')
```
Comment (sebLopezCot) [inline generator/Can_Library_h.py]: I think this should have a semicolon after this.

File: static_can.h [hunk 2]
```diff
@@ -1,28 +1,27 @@
-#ifndef _MY18_STATIC_CAN_H
-#define _MY18_STATIC_CAN_H
+#ifndef _CAN_LIBRARY_STATIC_CAN_H
+#define _CAN_LIBRARY_STATIC_CAN_H
 
 #include "can_raw.h"
 #include <stdint.h>
 #include <stdbool.h>
 
 Can_MsgID_T Can_MsgType(void);
```
Comment (sebLopezCot) [inline static_can.h]: We probably don't need this in the header file anymore

Comment (sebLopezCot) [reply static_can.h]: Can_MsgID_T Can_MsgType(void); that is

File: generator/Can_Library_h.py (hunk 1 above)
Comment (sebLopezCot) [reply generator/Can_Library_h.py]: since ...IDENTIFY is a function decl.

File: generator/Can_Library_h.py [hunk 3]
```diff
@@ -1,75 +1,72 @@
...
-
-        # Create enum
-        f.write("typedef enum {\n" +
-                "  Can_No_Msg,\n" +
-                "  Can_Unknown_Msg,\n" +
-                "  Can_Error_Msg,\n")
+        fw(
+            '#include <stdint.h>' '\n'
+            '#include <stdbool.h>' '\n\n'
+            '#include "static_can.h"' '\n\n'
```
Comment (sebLopezCot) [inline generator/Can_Library_h.py]: Also need to include "structs.h" here to include can0_T

File: generator/structs.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        fw('#include "enum_segments.h"\n\n')
+
+        for bus in car.buses:
+            for msg in bus.messages:
+                fw('typedef struct {\n')
+
+                for seg in msg.segments:
+                    if seg.c_type == 'enum':
+                        enum_name = coord(bus.name, msg.name, seg.name)
+                        fw('\t{} {};\n'.format(enum_name, seg.name))
```
Comment (sebLopezCot) [inline generator/structs.py]: Need to add an _T to the end of this format string (example: can0_VcuToDash_limp_state_T is correct, but this code actually only inputs can0_VcuToDash_limp_state)

File: generator/Can_Library_c.py [hunk 5]
```diff
@@ -52,121 +56,132 @@ def write(output_path, spec_path, base_path):
...
+                        fw(
+                            '\t' 'bitstring = INSERT(type_in->' + seg.name + ', bitstring, ' + str(seg.position) +
+                            ', ' + str(seg.length) + ');' '\n'
+                        )
+
+                length = max(seg.position + seg.length for seg in msg.segments)
+
+                fw(
+                    '\t' 'from_bitstring(&bitstring, can_out->data);' '\n'
+                    '\t' 'can_out->id = ' + msg.name.upper() + '__id;' '\n'
```
Comment (sebLopezCot) [inline generator/Can_Library_c.py]: msg.name.upper() no longer valid (ex: VCUTODASH__id is no longer valid here)

File: generator/Can_Library_c.py [hunk 6]
```diff
@@ -52,121 +56,132 @@ def write(output_path, spec_path, base_path):
...
+                    '\t' 'uint64_t bitstring = 0;' '\n'
+                    '\t' 'to_bitstring(can_in->data, &bitstring);\n'
+                )
+
+                for seg in msg.segments:
+                    if seg.c_type == 'enum':
+                        enum_name = 'Can_' + msg.name + 'ID_T'
+
+                        fw(
+                            '\t' 'type_out->' + seg.name + ' = (' + enum_name + ')EXTRACT(bitstring, ' +
```
Comment (sebLopezCot) [inline generator/Can_Library_c.py]: This fails for VCUTODASH__id

	type_out->limp_state = (Can_VcuToDashID_T)EXTRACT(bitstring, 50, 3);

Comment (sebLopezCot) [reply generator/Can_Library_c.py]: and some others

File: generator/Can_Library_c.py (hunk 5 above)
Comment (nistath) [reply generator/Can_Library_c.py]: When Ctrl+Shitf+F fails you.
Press F to pay respects.

**Comment by nistath** [comment]
Do you think that refactoring this further could be a freshman project next year? :P

**Comment by nistath** [comment]
@sebLopezCot, there is a potential naming bug lying in `bus.name_anything` so I chose caps just to reduce the probability of it happening. The conflict would exist if you had a message named `IDENTIFY` in any bus (conflicts with message form names formatted `bus.name_msg.name`).

How about renaming to `identify_bus.name` and `init_bus.name_board.name`?

**Comment by nistath** [comment]
Also, if we have `_` as the official delimiter, we might as well not have `_` in the spec names to avoid confusion. Unless there's an alternative.

File: generator/Can_Library_c.py [hunk 7]
```diff
@@ -1,42 +1,46 @@
-"""
+'''
 Generate Can_Libary.c file.
 Run this file (with the spec path as a command line argument) to write just
 Can_Libary.c or main.py to write all files.
-"""
+'''
 import sys
-sys.path.append("ParseCAN")
+sys.path.append('ParseCAN')
```
Comment (daniwhite) [inline generator/Can_Library_c.py]: Glad you're fixing this, I knew I was inconsistent on quotes but never got around to fixing them...

File: generator/Can_Library_c.py [hunk 8]
```diff
@@ -1,42 +1,46 @@
...
         for board in car.boards:
             if board.arch:  # Means it's a board we program
                 for bus in board.subscribe:
-                    f.write("void " + bus.name.title() + "_" + board.name.title() +
-                            "_Init(uint32_t baudrate) {\n")
-                    f.write("  Can_Init(baudrate);\n")
+                    fw('void init_{}(uint32_t baudrate) '.format(coord(bus.name, board.name)))
+                    fw(
+                        '{' '\n'
+                        '\t' 'Can_Init(baudrate);' '\n'
```
Comment (daniwhite) [inline generator/Can_Library_c.py]: Small thing: I thought the policy (at least for MY18 code) was spaces for indentation. Why the change?

File: generator/Can_Library_c.py [hunk 9]
```diff
@@ -52,121 +56,132 @@ def write(output_path, spec_path, base_path):
                     mask = mask ^ max_possible_id
 
                     # Set mask (pass in binary to make file more readable)
```
Comment (daniwhite) [inline generator/Can_Library_c.py]: Probably a good idea to update this comment so it matches the code.

File: generator/common.py [hunk 10]
```diff
@@ -1,6 +1,47 @@
...
+constants_path = '../constants.h'
+can_lib_c_path = '../Can_Library.c'
+can_lib_h_path = '../Can_Library.h'
+enum_segments_path = '../enum_segments.h'
+structs_path = '../structs.h'
 
-can_lib_c_base_path = "templates/Can_Library_BASE.c"
+can_lib_c_base_path = 'templates/Can_Library_BASE.c'
+
+def coord(*args):
```
Comment (daniwhite) [inline generator/common.py]: I definitely like the addition of this file. I think it also makes it easier for users of the library to figure out what names they should use based on the actual code and not just trusting the docs.

File: generator/Can_Library_c.py (hunk 8 above)
Comment (nistath) [reply generator/Can_Library_c.py]: CANlib is not MY18 code. :P It's always easy to find and replace '\t'. Maybe not so much for '    '. I picked tabs due to the nature of the project -- autogenerating code.

**Comment by sebLopezCot** [comment]
Things look good on stm32f2, should probs test on an lpc before approving...

**Comment by nistath** [comment]
LPC works with 2 CAN nodes. Minor changes coming before merge. @sebLopezCot

**Comment by sebLopezCot** [comment]
New changes look good and compile on the stm32f2, didn't test, but changes are minor. Don't know if the last commit is tested on LPC, but I think we're fine here. Nice work team!!!
